zilliqa/src/db.rs

```diff
@@ -1,234 +1,568 @@
+            ($cond: tt, $key: tt) => {
+                self.block_store.lock().unwrap().query_row(concat!("SELECT block_hash, view, height, parent_hash, signature, state_root_hash, timestamp, gas_used, qc, agg FROM blocks WHERE ", $cond), [$key], make_block).optional()?
+            };
+        }
+        Ok(match key {
+            Either::Left(hash) => {
+                query_block!("block_hash = ?1", hash)
+            }
+            Either::Right(view) => {
+                query_block!("view = ?1", view)
+            }
+        })
+    }
+
+    pub fn get_block(&self, key: Either<&Hash, &u64>) -> Result<Option<Block>> {
+        let Some(mut block) = self.get_transactionless_block(key)? else {
+            return Ok(None);
+        };
+        let transaction_hashes = self
+            .block_store
+            .lock()
+            .unwrap()
+            .prepare_cached("SELECT tx_hash FROM receipts WHERE block_hash = ?1")?
+            .query_map([block.header.hash], |row| row.get(0))?
+            .collect::<Result<Vec<Hash>, _>>()?;
+        block.transactions = transaction_hashes;
+        Ok(Some(block))
+    }
+
+    pub fn get_block_by_hash(&self, block_hash: &Hash) -> Result<Option<Block>> {
+        self.get_block(Either::Left(block_hash))
+    }
+
+    pub fn get_block_by_view(&self, view: &u64) -> Result<Option<Block>> {
+        self.get_block(Either::Right(view))
+    }
+
+    pub fn contains_block(&self, block_hash: &Hash) -> Result<bool> {
+        Ok(self
+            .block_store
+            .lock()
+            .unwrap()
+            .query_row(
+                "SELECT 1 FROM blocks WHERE block_hash = ?1",
+                [block_hash],
+                |row| row.get::<_, i64>(0),
+            )
+            .optional()?
+            .is_some())
+    }
+
+    fn make_receipt(row: &Row) -> rusqlite::Result<TransactionReceipt> {
+        Ok(TransactionReceipt {
+            tx_hash: row.get(0)?,
+            block_hash: row.get(1)?,
+            index: row.get(2)?,
+            success: row.get(3)?,
+            gas_used: row.get(4)?,
+            contract_address: row.get::<_, Option<AddressSqlable>>(5)?.map(|a| a.into()),
+            logs: row.get::<_, VecLogSqlable>(6)?.into(),
+            accepted: row.get(7)?,
+            errors: row.get::<_, MapScillaErrorSqlable>(8)?.into(),
+            exceptions: row.get::<_, VecScillaExceptionSqlable>(9)?.into(),
+        })
+    }
+
+    pub fn insert_transaction_receipt_with_db_tx(
+        &self,
+        sqlite_tx: &Connection,
+        receipt: TransactionReceipt,
+    ) -> Result<()> {
+        sqlite_tx.execute(
+            "INSERT INTO receipts
+                (tx_hash, block_hash, tx_index, success, gas_used, contract_address, logs, accepted, errors, exceptions)
+            VALUES (:tx_hash, :block_hash, :tx_index, :success, :gas_used, :contract_address, :logs, :accepted, :errors, :exceptions)
+            ON CONFLICT DO NOTHING",
```

Comment:
We do nothing - I added this because we have a habit of multiple-inserting transactions and receipts during our block execution flow. In `sled`, it'd be inserting the same data under the same key so it'd be a no-op, so for now I replicated that behaviour. A better solution would be adding guards to prevent double-inserts - and eventually refactoring consensus to eliminate them entirely - but I took the lazy path here, since it was already working this way with sled

Reply:
Actually, I thought I might have added some extra checks around where we insert transactions, so I went back to check and indeed this is no longer necessary. I added it partway through development before I had finished updating `consensus.rs`, and didn't realise it wasn't needed in the end.

Removed, all tests still pass, thanks for catching!